### PR TITLE
playing with sockets in kernel space

### PR DESCRIPTION
**Pull request by gszr**
Needs review

### CODE REVIEW

File: staging/sockets/luasockets.c [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+	lua_pushinteger(L, cnt);
+
+	return 1;
+}
+
+static int
+sockets_read(lua_State *L)
+{
+	struct luasocket* luasocket;
+	//XXX probably not a good idea; lua_pushlstring?
```
Comment (lneto) [inline staging/sockets/luasockets.c:157]: take a look at: 
[1] http://www.lua.org/manual/5.3/manual.html#luaL_Buffer
[2] https://github.com/openresty/lua-nginx-module/pull/519

File: staging/sockets/luasockets.c [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+	const char *type;
+	int sfd;
+	int err;
+	int dom = 0;
+	int typ = 0;
+
+	domain = luaL_checkstring(L, 1);
+	type   = luaL_checkstring(L, 2);
+
+	//XXX not about using strings to name domains (types, etc)
```
Comment (lneto) [inline staging/sockets/luasockets.c:37]: take a look at: http://w3.impa.br/~diego/software/luasocket/reference.html. Perhaps it's a better idea to mimic this API.

File: staging/sockets/luasockets.c [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+	lua_pushlstring(L, buff, cnt);
+
+	return 1;
+}
+
+static int
+luaopen_sockets(lua_State *L)
+{
+	const luaL_Reg sockets_lib[] = {
+		{"socket", sockets_socket},
```
Comment (lneto) [inline staging/sockets/luasockets.c:176]: I think new() is a better name.

File: staging/sockets/luasockets.c [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+	if ((luasocket = (struct luasocket *) lua_touserdata(L, 1)) == NULL)
+		luaL_error(L, "invalid Lua socket\n");
+	if ((err = soclose(luasocket->socket)) != 0)
+		luaL_error(L, "could not close socket (error %d)\n", err);
+	lua_pushboolean(L, 1);
+	return 1;
+}
+
+typedef int BindOrConnect(struct socket*, struct sockaddr*, struct lwp*);
```
Comment (lneto) [inline staging/sockets/luasockets.c:79]: would call that `op_t`, or something like that..

File: staging/sockets/luasockets.c [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+static int
+sockets_bind_connect(BindOrConnect handler, lua_State *L)
```
Comment (lneto) [inline staging/sockets/luasockets.c:82]: would call that `sockets_op(L, op)`

File: staging/sockets/luasockets.c [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+		lua_pushboolean(L, 1);
+	} else if (strcmp(domn, "local") == 0) {
+		;
+	}
+
+	return 1;
+}
+
+static int
```
Comment (lneto) [inline staging/sockets/luasockets.c:123]: inline

File: staging/sockets/luasockets.c [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+sockets_connect(lua_State *L)
+{
+	return sockets_bind_connect(soconnect, L);
+}
+
+static int
```
Comment (lneto) [inline staging/sockets/luasockets.c:129]: inline

File: staging/sockets/luasockets.c [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+		memset(&in4addr, 0, sizeof(in4addr));
+		in4addr.sin_len = sizeof(struct sockaddr_in);
+		in4addr.sin_addr.s_addr = inet_addr(addr);
+		in4addr.sin_port = htons(luaL_checkinteger(L, -1));
+		in4addr.sin_family = AF_INET;
+
+		solock(luasocket->socket);
+		if ((err = handler(luasocket->socket, (struct sockaddr*) &in4addr,
+			curlwp)) != 0)
+			luaL_error(L, "could not connect");
```
Comment (lneto) [inline staging/sockets/luasockets.c:112]: error message should be parametrized

File: staging/sockets/luasockets.c [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+	int err;
+
+	if ((luasocket = (struct luasocket *) lua_touserdata(L, 1)) == NULL)
+		luaL_error(L, "invalid socket");
+	domn = luaL_checkstring(L, 2);
+	addr = luaL_checkstring(L, 3);
+
+	// XXX inet4 and inet6 could share code
+	// XXX we could also declare a sockaddr in the function scope
+	// and then specialize it to sockaddr_in(,6, un) in the branches below
```
Comment (lneto) [inline staging/sockets/luasockets.c:96]: yes, we should modularize this in a better way.

File: staging/sockets/luasockets.c [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+#include <lauxlib.h>
+
+#include <stdio.h>
+
+MODULE(MODULE_CLASS_MISC, luasockets, "lua");
+
+#define BUF_SIZE 1024
+
+struct luasocket {
+	struct socket *socket;
```
Comment (lneto) [inline staging/sockets/luasockets.c:18]: for consistence, as you are using `fd` for file, I would use `so` to socket.

File: staging/sockets/luasockets.c [hunk 11]
```diff
@@ -0,0 +1,205 @@
+/* Lua sockets module */
```
Comment (lneto) [inline staging/sockets/luasockets.c:1]: I would use socket in the singular.

File: staging/sockets/luasockets.c [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+static int
+sockets_close(lua_State *L)
+{
+	struct luasocket *luasocket;
+	int err;
+
+	if ((luasocket = (struct luasocket *) lua_touserdata(L, 1)) == NULL)
+		luaL_error(L, "invalid Lua socket\n");
```
Comment (lneto) [inline staging/sockets/luasockets.c:72]: don't need `\n`, right?

File: staging/sockets/luasockets.c [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+	if ((err = soclose(luasocket->socket)) != 0)
+		luaL_error(L, "could not close socket (error %d)\n", err);
```
Comment (lneto) [inline staging/sockets/luasockets.c:74]: don't need `\n`, right?

File: staging/sockets/luasockets.c [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+
+	//XXX not about using strings to name domains (types, etc)
+	if (strcmp(domain, "inet4") == 0)
+		dom = AF_INET;
+	else if (strcmp(domain, "inet6"))
+		dom = AF_INET6;
+	else if (strcmp(domain, "local"))
+		dom = AF_LOCAL;
+	else
+		luaL_error(L, "'%s' domain not supported\n", domain);
```
Comment (lneto) [inline staging/sockets/luasockets.c:45]: don't need `\n`, right?

File: staging/sockets/luasockets.c [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+static int
+sockets_read(lua_State *L)
+{
+	struct luasocket* luasocket;
+	//XXX probably not a good idea; lua_pushlstring?
+	char buff[BUF_SIZE];
+	int cnt;
+
+	if ((luasocket = (struct luasocket*) lua_touserdata(L, 1)) == NULL)
+		luaL_error(L, "invalid socket");
```
Comment (lneto) [inline staging/sockets/luasockets.c:162]: you should also receive what you want to read, such as http://www.lua.org/manual/5.3/manual.html#pdf-file:read

**Comment by lneto** [comment]
@salazar, I think we shouldn't merge PRs before we have it reviewed; you can just keep working in a separate branch, then you can rebase/merge the reviewed PR in your work branch.

File: staging/sockets/luasockets.c (hunk 11 above)
Comment (gszr) [inline staging/sockets/luasockets.c:1]: Agreed & fixed

File: staging/sockets/luasockets.c (hunk 10 above)
Comment (gszr) [inline staging/sockets/luasockets.c:18]: Agreed & fixed

File: staging/sockets/luasockets.c (hunk 14 above)
Comment (gszr) [inline staging/sockets/luasockets.c:45]: oops ;x fixed

File: staging/sockets/luasockets.c (hunk 6 above)
Comment (gszr) [inline staging/sockets/luasockets.c:123]: oops ;x fixed

File: staging/sockets/luasockets.c (hunk 12 above)
Comment (gszr) [inline staging/sockets/luasockets.c:72]: oops ;x fixed

File: staging/sockets/luasockets.c (hunk 13 above)
Comment (gszr) [inline staging/sockets/luasockets.c:74]: oops ;x fixed

File: staging/sockets/luasockets.c (hunk 4 above)
Comment (gszr) [inline staging/sockets/luasockets.c:79]: Perhaps it'd be the case if we had more than 2 functions sharing the same interface? In this case I think it just makes it obscure -- it's easier to grasp what BindOrConnect means.

File: staging/sockets/luasockets.c (hunk 8 above)
Comment (gszr) [inline staging/sockets/luasockets.c:112]: Can you elaborate more on that? You mean printing the error code os something like that?

File: staging/sockets/luasockets.c (hunk 7 above)
Comment (gszr) [inline staging/sockets/luasockets.c:129]: Fixed

File: staging/sockets/luasockets.c (hunk 15 above)
Comment (gszr) [inline staging/sockets/luasockets.c:162]: Nice idea.

File: staging/sockets/luasockets.c (hunk 3 above)
Comment (gszr) [inline staging/sockets/luasockets.c:176]: Agreed & fixed

**Comment by gszr** [comment]
@lneto, Good point. In this case I merged before your review because we are working on a separate directory (staging) and it does not change a lot merging it or not.

Thanks for the tips :+1:

File: staging/sockets/luasockets.c (hunk 2 above)
Comment (gszr) [inline staging/sockets/luasockets.c:37]: I like the idea to split socket objects into different types, each one with a particular set of operations; e.g., socket.tcp() returns a 'tcp object', witch has a different set of operations from socket.udp(). That's something to keep in mind as a possibility.

File: staging/sockets/luasockets.c (hunk 1 above)
Comment (gszr) [inline staging/sockets/luasockets.c:157]: Checked & fixed. 

Btw, good job at fighting bloat (+17 −57) ; )

**Pull request merged**